server/datastore/mysql/operating_systems.go

```diff
@@ -0,0 +1,144 @@
+	os, err := maybeNewOperatingSystemDB(ctx, ds.writer, hostOS)
+	if err != nil {
+		return err
+	}
+	return maybeUpdateHostOperatingSystemDB(ctx, ds.writer, hostID, os.ID)
+}
+
+// maybeNewOperatingSystemDB queries the `operating_systems` table with the
+// name, version, arch, and kernel_version of the given operating system. If found,
+// it returns the record including the associated ID. If not found, it returns a call
+// to `newOperatingSystemDB`, which inserts a new record and returns the record
+// including the newly associated ID.
+func maybeNewOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostOS fleet.OperatingSystem) (*fleet.OperatingSystem, error) {
+	switch os, err := getOperatingSystemDB(ctx, tx, hostOS); {
+	case err == nil:
+		return os, nil
+	case errors.Is(err, sql.ErrNoRows):
+		return newOperatingSystemDB(ctx, tx, hostOS)
+	default:
+		return nil, ctxerr.Wrap(ctx, err, "get operating system")
+	}
+}
+
+// `newOperatingSystemDB` inserts a record for the given operating system and
+// returns the record including the newly associated ID.
+func newOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostOS fleet.OperatingSystem) (*fleet.OperatingSystem, error) {
+	stmt := "INSERT INTO operating_systems (name, version, arch, kernel_version) VALUES (?, ?, ?, ?)"
+	if _, err := tx.ExecContext(ctx, stmt, hostOS.Name, hostOS.Version, hostOS.Arch, hostOS.KernelVersion); err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "insert new operating system")
+	}
+
+	switch storedOS, err := getOperatingSystemDB(ctx, tx, hostOS); {
+	case err == nil:
+		return storedOS, nil
+	case errors.Is(err, sql.ErrNoRows):
+		return nil, doRetryErr
+	default:
+		return nil, ctxerr.Wrap(ctx, err, "get new operating system")
+	}
+}
+
+// getOperatingSystemDB queries the `operating_systems` table with the
+// name, version, arch, and kernel_version of the given operating system.
+// If found, it returns the record including the associated ID.
+func getOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostOS fleet.OperatingSystem) (*fleet.OperatingSystem, error) {
+	var os fleet.OperatingSystem
+	stmt := "SELECT * FROM operating_systems WHERE name = ? AND version = ? AND arch = ? AND kernel_version = ? LIMIT 1"
+	if err := sqlx.GetContext(ctx, tx, &os, stmt, hostOS.Name, hostOS.Version, hostOS.Arch, hostOS.KernelVersion); err != nil {
+		return nil, err
+	}
+	return &os, nil
+}
+
+// maybeUpdateHostOperatingSystemDB checks the `host_operating_system` table
+// for the given host ID. If found, it compares the given operating system ID
+// against the stored operating system ID and updates the table accordingly.
+// If not found, it inserts a new record for the given host id with the given
+// operating system id.
+func maybeUpdateHostOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostID uint, osID uint) error {
```

Comment:
This can be done in a single query using `INSERT ... ON DUPLICATE KEY UPDATE ...`.

Reply:
I had considered that but thought this might be a situation where we'd want to watch out for the scaling [gotchas](https://fleetdm.com/handbook/engineering#insert-on-duplicate-update) since most likely be updating rather than inserting to this table.

Reply:
Hmm, I still think this approach is slower, because you are having to perform 2 queries. First, you select and check if it exists, then update or insert. If we expect that most of the the time the row exists, then it should be faster to just optimistically do an update. If the values aren't actually changing then it's a noop. In addition, since we are using the `CLIENT_FOUND_ROWS` flag when connecting to mysql, we know if there was a row matched in the where clause. All we need to do is check `Result.RowsAffected()`. If greater than 0, then there was a row with the host id and we are finished. If equal to 0, then we need to do an insert.

Reply:
> Hmm, I still think this approach is slower

Yeah, I would assume the row is there and do the UPDATE. If that fails because there's no row to update, then you do the insert. So you just do an UPDATE most of the time, and it'll fail only once (the first time).

Reply:
@chiiph, @michalnicp: The "try to UPDATE first, then INSERT if no row to update" approach makes sense for something like host seen times where we know the row value has changed (time marches on). But I don't think it works very well in a case like this where the row likely exists but isn't changing (host operating systems don't change very often). In that case, the RowsAffected for the noop UPDATE will also be 0 so we can't use that to detect if we need to INSERT a non-existent row. 

As I see it, we'd either need to go with INSERT ... ON DUPLICATE KEY UPDATE or something like my original approach where we check first if nothing changed (which also tells us if the row exists) before doing either an update or insert.

Reply:
> But I don't think it works very well in a case like this where the row likely exists but isn't changing (host operating systems don't change very often). In that case, the RowsAffected for the noop UPDATE will also be 0 so we can't use that to detect if we need to INSERT a non-existent row.

This is not true. We are using the `CLIENT_FOUND_ROWS` flag which changes the behaviour of `Result.RowsAffected()`.

Reply:
So my comment above was based on what I was seeing as rows affected in my TablePlus. Thanks to @michalnicp's help, we figured out that the client isn't configured with the `CLIENT_FOUND_ROWS` flag.